File: sdk/js/src/explorer.ts
/** Blockscout (Etherscan-compatible) explorer REST client for GGCHAIN. */
export class Explorer {
  constructor(public readonly baseUrl: string = "https://explorer.gghyper.net") {}

  private async _api<T>(params: Record<string, string>): Promise<T> {
    const qs = new URLSearchParams(params).toString();
    const r = await fetch(`${this.baseUrl}/api?${qs}`);
    if (!r.ok) throw new Error(`Explorer ${r.status} ${r.statusText}`);
    const j = await r.json() as any;
    if (j.status === "0" && j.message !== "OK") return j.result as T;
    return j.result as T;
  }

  /** List of normal transactions for an address. */
  async txList(address: string, page = 1, offset = 50): Promise<any[]> {
    return await this._api({ module: "account", action: "txlist", address, page: String(page), offset: String(offset), sort: "desc" });
  }

  /** List of internal transactions for an address. */
  async internalTxList(address: string): Promise<any[]> {
    return await this._api({ module: "account", action: "txlistinternal", address });
  }

  /** ERC-20 token transfers for an address. */
  async tokenTransfers(address: string, contractaddress?: string): Promise<any[]> {
    const p: Record<string, string> = { module: "account", action: "tokentx", address };
    if (contractaddress) p.contractaddress = contractaddress;
    return await this._api(p);
  }

  /** Get verified contract source code. */
  async getContractSource(address: string): Promise<any> {
    return await this._api({ module: "contract", action: "getsourcecode", address });
  }

  /** Get contract ABI (if verified). */
  async getContractABI(address: string): Promise<any[]> {
    const raw = await this._api<string>({ module: "contract", action: "getabi", address });
    return typeof raw === "string" ? JSON.parse(raw) : raw as any;
  }

  /** Token holder count + supply. */
  async tokenInfo(contractaddress: string): Promise<any> {
    return await this._api({ module: "token", action: "getToken", contractaddress });
  }

  /** Direct URL to an address or tx on the explorer UI. */
  url(addressOrTx: string): string {
    return addressOrTx.length === 42 ? `${this.baseUrl}/address/${addressOrTx}` : `${this.baseUrl}/tx/${addressOrTx}`;
  }
}

File: sdk/python/gghyper_sdk/explorer.py
"""GGCHAIN Blockscout explorer API client."""
from __future__ import annotations
from typing import Any, Optional
import requests
import json


class Explorer:
    def __init__(self, base_url: str = "https://explorer.gghyper.net"):
        self.base_url = base_url.rstrip("/")

    def _api(self, **params) -> Any:
        r = requests.get(f"{self.base_url}/api", params=params, timeout=15)
        r.raise_for_status()
        j = r.json()
        return j.get("result")

    def tx_list(self, address: str, page: int = 1, offset: int = 50):
        return self._api(module="account", action="txlist", address=address,
                         page=page, offset=offset, sort="desc")

    def internal_tx_list(self, address: str):
        return self._api(module="account", action="txlistinternal", address=address)

    def token_transfers(self, address: str, contractaddress: Optional[str] = None):
        p = {"module": "account", "action": "tokentx", "address": address}
        if contractaddress: p["contractaddress"] = contractaddress
        return self._api(**p)

    def get_contract_source(self, address: str) -> Any:
        return self._api(module="contract", action="getsourcecode", address=address)

    def get_contract_abi(self, address: str):
        raw = self._api(module="contract", action="getabi", address=address)
        return json.loads(raw) if isinstance(raw, str) else raw

    def token_info(self, contractaddress: str):
        return self._api(module="token", action="getToken", contractaddress=contractaddress)

    def url(self, address_or_tx: str) -> str:
        return f"{self.base_url}/{'address' if len(address_or_tx) == 42 else 'tx'}/{address_or_tx}"
